README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3243425.svg)](https://doi.org/10.5281/zenodo.3243425)


# bromeliad 1.0-3

A data package with distribution data for 3,272 bromeliaceae species across tropical America.

## Included data
The package includes five data files:

1. `taxonomy traits` - a data.frame with taxonomy, traits and conservation assessment for the included species.
2. `all_ranges` - a simple features object containing range polygons for 3,272 bromeliaceae species.
3. `model_binary` - a list of presence/absence rasters, for 541 species for which the ranges were estimated based on climate suitability models
4. `model_ensemble` - a list of rasters with habitat suitability, for 541 species for which the ranges were estimated based on climate suitability models
5. `digitized_records` - a data.frame with 1,207 point locations of 920 species type specimens, georeferenced from the literature for this package

## Installation
The package can be installed from github using `devtools::install_github`
```{r}
library(devtools)

install_github("idiv-biodiversity/bromeliad")

library(bromeliad)
```

## Documentation
The pacakge also includes some convinieince functions to select taxa and visualize species ranges and diversity. A vignette describing how to select individual taxa and visualize species distributions and diversity patterns is available [here](https://idiv-biodiversity.github.io/bromeliad/), the corresponding paper [here](https://onlinelibrary.wiley.com/doi/full/10.1111/ddi.13004).

## Contribute
If you want to contribute with new species or corrections, please contact the [maintainer](mailto:alexander.zizka@idiv.de) of the package.

## Licence
The package and data are available under a CC-BY license.

## Citation
If you use the data, please cite the package as Zizka A, Azevedo J, Leme E, NEves B, Costa AFC, Caceres D & Zizka, G (2019) Biogeography and conservation status of the pineapple family (Bromeliaceae). *Diversity and Distributions*, 26(2):183-195. DOI:10.1111/ddi.13004.


NEWS.md
bromeliad 1.0-3 (2020-05-14)
=========================

### MINOR IMPROVEMENTS
  * updated citation
  * updated license



bromeliad 1.0-2 (2019-11-25)
=========================

### MINOR IMPROVEMENTS
  * added a link to the webpage to the Readme 
  * improved the package description



R/data.R

#' taxonomy_traits
#'
#' A datset of species known to be phylogenetically derived woody
#'
#' @format A data frame with 17 variables:
#' \describe{
#' \item{\code{tax_accepted_name}}{the accepted species name}
#' \item{\code{tax_genus}}{the genus}
#' \item{\code{tax_tribus}}{the tribus}
#' \item{\code{tax_subfamily}}{the subfamily}
#' \item{\code{geo_range}}{the range size of the species, based on the modelled distribution}
#' \item{\code{conv_conr_eoo}}{the extent of occurrence}
#' \item{\code{conv_conr_aoo}}{the area of occupancy}
#' \item{\code{conv_conr_pop}}{the number of independent populations deived during automated assessment}
#' \item{\code{conv_conr_code}}{The detailled code of the conservation status based on automated assessment}
#' \item{\code{conv_conr_stat}}{The conservation status of the species based on automated conservation assessment}
#' \item{\code{tr_growth_form}}{the growth form of the species from the world checklist of selected plant families}
#' \item{\code{tax_canonical}}{the canonical species name}
#' }
#'
#'
"taxonomy_traits"


#' model_ensemble
#'
#' A list of raster layers, including the habitat suitability for 541 species as defined by ensemble modelling.
#'
#'
"model_ensemble"


#' model_binary
#'
#'
#' A list of raster layers, including the projected binary distributions for 541 species as defined by ensemble modelling.
#'
#'
"model_binary"


#' all ranges
#'
#' Polygons of the geographic distribution for 3295 bromeliad species as derived from distribution modelling.
#' For species with more than 9 occurrence records available modelled using climate SDMs,
#' for species with 3-8 records available modelled by convex hulls, and
#' for species with less than 3 records modelled by a geographic buffer with 50 kilometer radius
#'
#' @format Simple feature collection with 6 features and 1 field
#' geometry type:  GEOMETRY
#' dimension:      XY
#' bbox:           xmin: -78.6167 ymin: -30.27966 xmax: -36.88333 ymax: 0.3333
#' epsg (SRID):    4326
#' proj4string:    +proj=longlat +datum=WGS84 +no_defs
#' \describe{
#' \item{\code{tax_accepted_name}}{the scientific species name}
#' \item{\code{geometry}}{the range polygon of the species}
#' }
#'
#'
"all_ranges"

#' georef_additions
#'
#'
#' A data.frame of occurrence records derived from georeferencing type specimen from the literature.
#'
#' \describe{
#' \item{\code{accepted_name_species}}{}
#' \item{\code{scientificname}}{}
#' \item{\code{family}}{}
#' \item{\code{genus}}{}
#' \item{\code{species}}{}
#' \item{\code{taxonrank}}{}
#' \item{\code{decimallongitude}}{}
#' \item{\code{decimallatitude}}{}
#' \item{\code{coordinate_accuracy}}{}
#' \item{\code{flag}}{}
#' \item{\code{remarks}}{}
#' \item{\code{basisofrecords}}{}
#' \item{\code{source}}{}
#' }
"georef_additions"



R/get_range.R
#' Select Bromeliaceae Species Ranges by Taxonomy and Traits
#'
#'Get the geographic range for all species selected via the arguments.
#'The type of range estimate depends on the \dQuote{type} argument.
#'
#'Modelled ranges are available for 542 species,
#'range polygons for 2395 species. For species with model distribution, the range polygons are based on the models, otherwise
#'on a convex hull around available occurrence records, or a 50 km buffer for species with less than 3 occurrence records available.
#'See Zizka et al 2019 for methods.
#'
#' @param scientific a character vector of full scientific names including authorities, of the species of interest
#' @param canonical a character vector of canonical names, of the species of interest.
#' @param genus a character vector of genera names to select.
#' @param subfamily a character vector of subfamily names to select.
#' @param life_form a character vector of life forms to select.
#' @param assessment a character vector of conservation assessment to select.
#' @param range_size a vector of two numericals with the minimum and maximum range size (in square kilometres), to select.
#' @param type a cahracter defining the output format, see details
#'
#' @return Depending on the \dQuote{type} argument. If \dQuote{binary} a presence/absence raster based on the modelled habitat suitability,
#'  at 100x100 km resolution in Behrmann projection,
#'  if \dQuote{suitability} the habitat suitability as predicted by an ensemble model in Behrmann projection, and
#'  if {polygon} a simple feature object in lat/lon projection.
#' .
#' @examples
#' get_range(scientific = "Aechmea mexicana Baker")
#' get_range(scientific = "Aechmea mexicana Baker", type = "binary")
#' get_range(scientific = "Aechmea mexicana Baker", type = "suitability")
#' get_range(canonical = "Aechmea mexicana")
#' get_range(genus = "Aechmea")
#' get_range(genus = "Aechmea", type = "binary")
#' get_range(genus = c("Aechmea", "Zizkaea"))
#' get_range(subfamily = "Pitcairnioideae")
#' get_range(life_form = "epiphyte")
#' get_range(assessment = c("CR", "VU"))
#' get_range(range_size = c(1000, 10000))
#'
#' @export
#' @importFrom dplyr filter
#' @importFrom methods as
#' @importFrom rlang .data
#'
#'
get_range <- function(scientific = NULL, canonical = NULL,
                      genus = NULL, subfamily = NULL,
                      life_form = NULL,
                      assessment = NULL, range_size = NULL,
                      type = "polygon"){

  match.arg(type, choices = c("binary", "suitability", "polygon"))

  taxonomy_traits <- bromeliad::taxonomy_traits

  if(!is.null(scientific)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$tax_accepted_name %in% scientific)
  }

  if(!is.null(canonical)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$tax_canonical %in% canonical)
  }

  if(!is.null(genus)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$tax_genus %in% genus)
  }

  if(!is.null(subfamily)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$tax_subfamily %in% subfamily)
  }

  if(!is.null(life_form)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$tr_growth_form %in% life_form)
  }

  if(!is.null(assessment)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$conv_conr_stat %in% assessment)
  }

  if(!is.null(range_size)){
    taxonomy_traits <- dplyr::filter(taxonomy_traits, .data$geo_range > range_size[1] & .data$geo_range < range_size[2])
  }

  switch(type,
         binary = {model_binary <-  bromeliad::model_binary;
                   out <- model_binary[which(names(model_binary) %in% taxonomy_traits$tax_accepted_name)]},
         suitability = {model_ensemble <- bromeliad::model_ensemble; out <- model_ensemble[which(names(model_ensemble) %in% taxonomy_traits$tax_accepted_name)]},
         polygon = {all_ranges <- bromeliad::all_ranges; out <- dplyr::filter(all_ranges, .data$tax_accepted_name %in% taxonomy_traits$tax_accepted_name)}
           )

  class(out) <- c("bromeli", class(out))

  return(out)
}


R/map_richness.R
#' Mapping Bromeliad Species Richness
#'
#' Produces a richness map in a 100x100 km grid for the selected bromeliaceae species.
#'
#' @param x an object of the class \dQuote{bromeli} generated by get_range.
#' @param title An optional title to be added to the plot.
#'
#' @return aa richness map of the selected taxa across the Neotropics.
#' .
#' @examples
#' map_richness(x = get_range(genus = "Aechmea", type = "binary"))
#'
#' @export
#' @importFrom rnaturalearth ne_download
#' @importFrom speciesgeocodeR RangeRichness
#' @importFrom sp spTransform CRS
#' @importFrom sf st_transform
#' @importFrom ggplot2 aes annotate fortify ggtitle coord_fixed fortify geom_polygon geom_sf geom_tile ggplot theme xlim ylim
#' @importFrom ggthemes theme_map
#' @importFrom viridis scale_fill_viridis
#' @importFrom raster crop extend extent raster rasterToPoints



map_richness <- function(x, title = NULL){

  if(!"bromeli" %in% class(x)){
      stop("x must be of class bromeli, generated by 'get_range'")
  }else{
    class(x) <- class(x)[!class(x) == "bromeli"]
  }
  # Projection
  wgs1984 <- "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
  behr <- '+proj=cea +lon_0=0 +lat_ts=30 +x_0=0 +y_0=0 +datum=WGS84 +ellps=WGS84 +units=m +no_defs'

  if("sf" %in% class(x)){
    dat <- sf::st_transform(x, behr)

    ## transfom to sp
    dat <- as(dat, 'Spatial')

    # create dummy raster
    be <- raster::raster(ncol = 360, nrow = 142, xmn = -17367529, xmx = 17367529, ymn = -6356742,
                 ymx = 7348382, crs = behr)
    be <- raster::crop(be, raster::extent(dat))

    # get richness raster
    ri <- speciesgeocodeR::RangeRichness(dat, ras = be)
  }else{
    # standardize extent
    be <- raster::raster(ncol = 360, nrow = 142, xmn = -17367529, xmx = 17367529, ymn = -6356742,
                 ymx = 7348382, crs = behr)
    e <- raster::extent(be)
    re <-  lapply(x, function(r){raster::extend(r,e)})

    out <-  list()
    re <- for(i in 1:length(re)){
      sub <- re[[i]]
      sub[is.na(sub)] <- 0
      out[[i]] <-  sub
      }

    # get richness raster
    ri <- Reduce("+", out)
  }


  #Background map
  world.inp  <- rnaturalearth::ne_download(scale = 110,
                                           type = 'countries',
                                           load = TRUE)

  world.behr <- sp::spTransform(world.inp, sp::CRS(behr)) %>% ggplot2::fortify()

  #convert to ggplot2 format
  ri <- data.frame(raster::rasterToPoints(ri))%>%
    mutate(layer = ifelse(.data$layer == 0, NA, .data$layer))

  # Plot
  if(sum(ri$layer,na.rm = TRUE) > 0){
    plo_tot <- ggplot()+
      ggplot2::geom_tile(data = ri, aes(x = .data$x, y = .data$y, fill = .data$layer), alpha = 1)+
      viridis::scale_fill_viridis(name = "Species", direction = -1, na.value = "transparent")+
      geom_polygon(data = world.behr,
                   aes(x = .data$long, y = .data$lat, group = .data$group), fill = "transparent", color = "black")+
      ggplot2::xlim(-12000000, -3000000)+
      ggplot2::ylim(-6500000, 4500000)+
      ggplot2::coord_fixed()+
      ggthemes::theme_map()+
      ggplot2::theme(legend.justification = c(0, 0),
                     legend.position = c(0, 0))
  }else{
    plo_tot <- ggplot()+
      ggplot2::geom_tile(data = ri, aes(x = .data$x, y = .data$y), alpha = 1)+
     # viridis::scale_fill_viridis(name = "Species", direction = -1, na.value = "transparent")+
      geom_polygon(data = world.behr,
                   aes(x = .data$long, y = .data$lat, group = .data$group), fill = "transparent", color = "black")+
      ggplot2::xlim(-12000000, -3000000)+
      ggplot2::ylim(-6500000, 4500000)+
      ggplot2::coord_fixed()+
      ggthemes::theme_map()+
      ggplot2::theme(legend.justification = c(0, 0),
                     legend.position = c(0, 0))
  }


  if(!is.null(title)){
    plo_tot <- plo_tot+
      ggtitle(title)
  }

  plo_tot
}


prepare_datasets.R
library(bromeliad)
library(raster)
library(tidyverse)
library(sf)
library(stringi)

#load data
load(file = "C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/bromeliaceae_ranges_with_data.rda")

# all ranges
all_ranges <- bromeli %>%
  select(tax_accepted_name)

save(all_ranges, file = "data/all_ranges.rda", compress = "xz")

# taxonomy_traits
head(taxonomy_traits)

st_geometry(bromeli) <- NULL

taxonomy_traits <- bromeli %>%
  select(tax_accepted_name, tax_genus, tax_tribus, tax_subfamily,
         geo_range, conv_conr_eoo, conv_conr_aoo, conv_conr_pop, conv_conr_code, conv_conr_stat, tr_growth_form) %>%
  mutate(tax_canonical = paste(str_split_fixed(tax_accepted_name, n = 3, pattern = " ")[,1],
                               str_split_fixed(tax_accepted_name, n = 3, pattern = " ")[,2],
                               sep = " ")) %>%
  as.tibble()


save(taxonomy_traits, file = "data/taxonomy_traits.rda", compress = "xz")

# Ensemble models
head(model_ensemble)

lis <- list.files("C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/ensemble_rasters")

modens <- lapply(paste("C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/ensemble_rasters/",
                       lis,
                       sep = ""), "raster")
modens <- lapply(modens, "readAll")
model_ensemble <- modens
names(model_ensemble) <- gsub(".asc", "", lis)
save(model_ensemble, file = "data/model_ensemble.rda", compress = "xz")

# binary models
head(model_binary)

lis <- list.files("C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/binary_rasters")

modens <- lapply(paste("C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/binary_rasters/",
                      lis,
                      sep = ""), "raster")
modens <- lapply(modens, "readAll")
model_binary <- modens
names(model_binary) <- gsub(".asc", "", lis)
save(model_binary, file = "data/model_binary.rda", compress = "xz")


head(georef_additions)
load("C:/Users/az64mycy/Dropbox (iDiv)/research_projects/31_bromeliaceae_distribution/bromeliaceae_distributions/output/manual_georeferencing_all.rds")

georef_additions <- out_all
save(georef_additions , file = "data/georef_additions.rda", compress = "xz")


R/map_species.R
#' Mapping species distributions
#'
#' Produces distribution maps for all species in the selection
#'
#' Modelled ranges are available for 542 species,
#' range polygons for 2395 species. For species with model distribution, the range polygons are based on the models, otherwise
#' on a convex hull around available occurrence records, or a 50 km buffer for species with less than 3 occurrence records available.
#' See Zizka et al 2019 for methods.
#'
#' @param model character string. If \dQuote{binary} presence absence per 100 km grid cell is plotted,
#' if \dQuote{suitability} the modelled habitat suitability is plotted.
#' @param polygon logical. If TRUE, an outline of the range polygon is plotted.
#' @param iucn logical. If TRUE a legend with the conservation assessment is added to the plot.
#' @inheritParams get_range
#'
#' @return a plot for each selected species
#' .
#' @examples
#' map_species(canonical = "Aechmea mexicana")
#'
#' @export
#' @importFrom magrittr %>%
#' @importFrom dplyr filter mutate
#' @importFrom rnaturalearth ne_download
#' @importFrom sp spTransform CRS
#' @importFrom ggplot2 aes annotate element_line ggtitle coord_fixed fortify geom_polygon geom_sf geom_tile ggplot theme xlim ylim
#' @importFrom ggthemes theme_map
#' @importFrom sf st_transform
#' @importFrom viridis scale_fill_viridis
#' @importFrom raster rasterToPoints
#' @importFrom utils data

map_species <-  function(scientific = NULL, canonical = NULL,
                         genus = NULL, subfamily = NULL,
                         life_form = NULL,
                         assessment = NULL, range_size = NULL,
                         model = "suitability",
                         polygon = TRUE, iucn = TRUE){


  match.arg(model, choices = c("binary", "suitability"))

  # projetions
  wgs1984 <- "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
  behr <- '+proj=cea +lon_0=0 +lat_ts=30 +x_0=0 +y_0=0 +datum=WGS84 +ellps=WGS84 +units=m +no_defs'

  taxonomy_traits <-  bromeliad::taxonomy_traits

  if(polygon){
    #data("all_ranges", envir = environment())
    rang <- get_range(scientific = scientific, canonical = canonical,
                      genus = genus, subfamily = subfamily,
                      life_form = life_form,
                      assessment = assessment, range_size = range_size)
    rang <- st_transform(rang, crs = behr)
  }else{
    rang <- NULL
  }

  if(model == "suitability"){
    #data("model_ensemble", envir = environment())
    suit <- get_range(scientific = scientific, canonical = canonical,
                      genus = genus, subfamily = subfamily,
                      life_form = life_form,
                      assessment = assessment, range_size = range_size,
                      type = "suitability")
  }else{
    suit <- NULL
  }

  if(model == "binary"){
    #data("model_binary", envir = environment())
    bina <- get_range(scientific = scientific, canonical = canonical,
                      genus = genus, subfamily = subfamily,
                      life_form = life_form,
                      assessment = assessment, range_size = range_size,
                      type = "binary")
  }else{
    bina <- NULL
  }

  #background map
  world.inp  <- rnaturalearth::ne_download(scale = 50,
                                           type = 'land',
                                           category = 'physical',
                                           load = TRUE)

  world.behr <- sp::spTransform(world.inp, sp::CRS(behr)) %>% fortify()

  # Background map
  plo_bg <- ggplot()+
    # geom_tile(data = plo_tot, aes(x = x, y = y, fill = layer), alpha = 0.8)+
    # scale_fill_viridis(name = "Species", direction = -1, na.value = "transparent")+
    geom_polygon(data = world.behr,
                 ggplot2::aes(x = .data$long, y = .data$lat, group = .data$group),
                 fill = "transparent", color = "transparent")+
    xlim(-12000000, -3000000)+
    ylim(-6500000, 4500000)+
    coord_fixed()+
    # ggtitle("Bromeliaceae")+
    theme_map()+
    theme(legend.justification = c(0, 0),
          legend.position = c(0, 0))

  if(polygon){
    sps <- unique(rang$tax_accepted_name)

    if(!is.null(bina)){
      mod_li <- names(bina)
    }
    if(!is.null(suit)){
      mod_li <- names(suit)
    }
  }else{
    if(!is.null(bina)){
      sps <- unique(names(bina))
      mod_li <- names(bina)
    }
    if(!is.null(suit)){
      sps <- unique(names(suit))
      mod_li <- names(suit)
    }
  }

  #A loop to produce one map per species and page to
  for(i in 1:length(sps)){
    #title
    plo.subs <- plo_bg+
      ggtitle(sps[i])

    # modelled_distribution
    if(!is.null(bina)){
      mod_sub <- bina[names(bina) == sps[i]]
      mod_sub <-  data.frame(rasterToPoints(mod_sub[[1]]))
      plo.subs <- plo.subs+
        geom_tile(data = mod_sub,
                  ggplot2::aes(x = .data$x, y = .data$y),
                  alpha = 0.8,
                  color = "darkgreen")+
        geom_polygon(data = world.behr,
                     ggplot2::aes(x = .data$long, y = .data$lat, group = .data$group),
                     fill = "transparent",
                     color = "black")+
        theme(legend.position = c(1, 0.1))
    }

    if(!is.null(suit)){
      mod_sub <- suit[names(suit) == sps[i]]
      mod_sub <-  data.frame(rasterToPoints(mod_sub[[1]]))
      names(mod_sub) <- c("x", "y", "layer")
      plo.subs <- plo.subs+
          geom_tile(data = mod_sub,
                    ggplot2::aes(x = .data$x, y = .data$y, fill = .data$layer),
                    alpha = 0.8)+
          scale_fill_viridis(name = "Habitat\nsuitability",
                             direction = -1,
                             na.value = "transparent")+
        geom_polygon(data = world.behr,
                     ggplot2::aes(x = .data$long, y = .data$lat, group = .data$group),
                     fill = "transparent",
                     color = "black")+
        theme(legend.position = c(1, 0.1))
    }

    #range polygons
    if(!is.null(rang)){
      rang.sub <-  filter(rang, .data$tax_accepted_name == sps[i])

      plo.subs <- plo.subs+
        geom_sf(data = rang.sub,  alpha = 0, colour = "red", datum = NA, fill = "transparent", lwd = 0.5)+
        theme(panel.grid.major = element_line(colour = 'transparent'))
    }

    #Conservation statues
    if(iucn){
      rang.sub <-  filter(taxonomy_traits, .data$tax_accepted_name == sps[i])

      plo.subs <- plo.subs+
        annotate("text",
                 x = -11481867,
                 y = -3698674,
                 label = rang.sub$conv_conr_code,
                 adj = 0,
                 fontface =2,
                 size = 5)+
        annotate("text",
                 x = rep(-11481867, 3),
                 y = c(-4198674, -4707084, -5180102),
                 label = c("EOO:", "AOO:", "Subpop:"),
                 adj = 0)+
        annotate("text", x = rep(-9066197, 3),
                 y = c(-4198674, -4707084, -5180102),
                 label = c(rang.sub$conv_conr_eoo, rang.sub$conv_conr_aoo, rang.sub$conv_conr_pop),
                 adj = 1)
    }


    # Output
    print(plo.subs)
  }


}
